src/trace/context/extractor.ts

```diff
@@ -0,0 +1,122 @@
+import { Context } from "aws-lambda";
+import { TraceConfig } from "../listener";
+import { logDebug, logError } from "../../utils";
+import { XrayService } from "../xray-service";
+import {
+  AppSyncEventTraceExtractor,
+  CustomTraceExtractor,
+  EventBridgeEventTraceExtractor,
+  EventBridgeSQSEventTraceExtractor,
+  HTTPEventTraceExtractor,
+  KinesisEventTraceExtractor,
+  LambdaContextTraceExtractor,
+  SNSEventTraceExtractor,
+  SNSSQSEventTraceExtractor,
+  SQSEventTraceExtractor,
+  StepFunctionEventTraceExtractor,
+} from "./extractors";
+import { StepFunctionContextService } from "../step-function-service";
+import { EventValidator } from "../../utils/event-validator";
+import { TracerWrapper } from "../tracer-wrapper";
+import { SpanContextWrapper } from "trace/span-context-wrapper";
+
+export const DATADOG_TRACE_ID_HEADER = "x-datadog-trace-id";
+export const DATADOG_PARENT_ID_HEADER = "x-datadog-parent-id";
+export const DATADOG_SAMPLING_PRIORITY_HEADER = "x-datadog-sampling-priority";
+
+export interface EventTraceExtractor {
+  extract(event: any): SpanContextWrapper | null;
+}
+
+export interface TraceHeaders {
+  [DATADOG_TRACE_ID_HEADER]: string;
+  [DATADOG_PARENT_ID_HEADER]: string;
+  [DATADOG_SAMPLING_PRIORITY_HEADER]: string;
+}
+
+export class TraceContextExtractor {
+  private xrayService: XrayService;
+  private stepFunctionContextService?: StepFunctionContextService;
+
+  constructor(private tracerWrapper: TracerWrapper, private config: TraceConfig) {
+    this.xrayService = new XrayService();
+  }
+
+  async extract(event: any, context: Context): Promise<SpanContextWrapper | null> {
+    this.stepFunctionContextService = StepFunctionContextService.instance(event);
+
+    let spanContext: SpanContextWrapper | null = null;
+    if (this.config.traceExtractor) {
+      const customExtractor = new CustomTraceExtractor(this.config.traceExtractor);
+      spanContext = await customExtractor.extract(event, context);
+    }
+
+    if (spanContext === null) {
```

Comment:
Mmm, my current thought would be doing early returns, but this would imply to use the `this.addTraceContextToXray` method three times instead of one. That way, we avoid `null` checking, and we ensure early returns, wdyt?

Reply:
Yeah I realize now you're copying this over anyway so let's leave that for later.